gs_chunked_io/async_collections.py

```diff
@@ -0,0 +1,125 @@
+"""
+Set-like and list-like collections to manage concurrent execution using ThreadPoolExecutor.
+"""
+import multiprocessing
+from concurrent.futures import Future, ThreadPoolExecutor, as_completed
+from typing import Any, List, Set, Callable, Generator, Optional
+
+
+class _AsyncCollection:
+    def __init__(self, executor: ThreadPoolExecutor, size: int=0):
```

Comment:
Nit: It's not immediately clear to me what size is.  Maybe consider "threads"?

Should the default also be `multiprocessing.cpu_count()`, since the current default will cause an error?

Reply:
I'll try to think of a better name than "size". It reflects the "maxsize" parameter of Python's [queue](https://docs.python.org/3/library/queue.html), but it's not quite right.

This is a base class that should not be instantiated directly. Derived classes are expected to provide defaults for `size`.

Reply:
Changed `size` -> `concurrency`